SharedWorkerGlobalScope.close() does not then allow the SharedWorker to be replaced

cdumez opened the pull request:
#### 76527da41c8016eef5fc6f722377cd71f4d7ce47
<pre>
SharedWorkerGlobalScope.close() does not then allow the SharedWorker to be replaced
<a href="https://bugs.webkit.org/show_bug.cgi?id=250839">https://bugs.webkit.org/show_bug.cgi?id=250839</a>
rdar://problem/104481929

Reviewed by Youenn Fablet.

Shared workers are usually terminated once all their clients go away and the
NetworkProcess is in charge of this.

However, the shared worker can also choose to terminate itself by calling
`self.close()`. In this case, we were failing to notify the network process
that the shared worker had terminated. As a result, when a shared worker
connection request would come in later on, the network process would still
direct it towards the existing (now terminated) shared worker.

To address this issue, we now IPC back from the shared worker process to the
network process when a shared worker gets terminated. This allows the network
process state to remain up-to-date and to make more accurate decisions.

However, doing only this would still be flaky since there can be a race between
the NetworkProcess receiving the IPC about a shared worker being terminated
and a new connection request received by the NetworkProcess. To address this,
I now added a CompletionHandler to the PostConnectEvent IPC from the
NetworkProcess to the SharedWorker process, indicating if we managed to connect
this SharedWorker object to the existing SharedWorker context. In case of
failure, we clean the state in the NetworkProcess and re-attempt the connection,
which will relaunch a SharedWorker.

* LayoutTests/http/wpt/shared-workers/relaunch-after-close-expected.txt: Added.
* LayoutTests/http/wpt/shared-workers/relaunch-after-close.html: Added.
* LayoutTests/http/wpt/shared-workers/resources/relaunch-after-close-page2.html: Added.
* LayoutTests/http/wpt/shared-workers/resources/relaunch-after-close-sharedworker.js: Added.
(shutdown):
(broadcast):
(sendMessageToPortRef):
* LayoutTests/http/wpt/shared-workers/resources/relaunch-after-close-window.js: Added.
(testComplete):
(async beginTestForPage1):
(async beginTestForPage2):
(async initSharedWorker):
(disonnectFromSharedWorker):
(async tellSharedWorkerToShutdown):
(log):
(deferredWithTimeout):
(async wait):
* Source/WebCore/workers/shared/context/SharedWorkerContextManager.cpp:
(WebCore::SharedWorkerContextManager::stopSharedWorker):
* Source/WebCore/workers/shared/context/SharedWorkerContextManager.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::sharedWorkerTerminated):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.cpp:
(WebKit::WebSharedWorkerServerToContextConnection::sharedWorkerTerminated):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::sharedWorkerTerminated):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.h:

Canonical link: <a href="https://commits.webkit.org/259228@main">https://commits.webkit.org/259228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fb4debf8e0f5680b42d3a4eb390a1e752aa5c2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104385 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113602 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173896 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14565 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4380 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112642 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110152 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94291 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93095 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25902 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6825 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27259 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6955 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3807 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12981 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46817 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6361 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8749 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/workers/shared/context/SharedWorkerContextManager.cpp:
```diff
@@ -57,7 +57,10 @@ void SharedWorkerContextManager::stopSharedWorker(SharedWorkerIdentifier sharedW
-    thread.stop([worker = WTFMove(worker)]() mutable {
+    thread.stop([worker = WTFMove(worker), sharedWorkerIdentifier]() mutable {
+        if (auto* connection = SharedWorkerContextManager::singleton().connection())
+            connection->sharedWorkerTerminated(sharedWorkerIdentifier);
```
Do we need to do it here?
It seems we should call sharedWorkerTerminated as soon as we remove the worker from the map?

cdumez replied:
I matched the corresponding service worker code iirc but you may be right. I'll check.

cdumez committed:
SharedWorkerGlobalScope.close() does not then allow the SharedWorker to be replaced
https://bugs.webkit.org/show_bug.cgi?id=250839
rdar://problem/104481929

Reviewed by Youenn Fablet.

Shared workers are usually terminated once all their clients go away and the
NetworkProcess is in charge of this.

However, the shared worker can also choose to terminate itself by calling
`self.close()`. In this case, we were failing to notify the network process
that the shared worker had terminated. As a result, when a shared worker
connection request would come in later on, the network process would still
direct it towards the existing (now terminated) shared worker.

To address this issue, we now IPC back from the shared worker process to the
network process when a shared worker gets terminated. This allows the network
process state to remain up-to-date and to make more accurate decisions.

However, doing only this would still be flaky since there can be a race between
the NetworkProcess receiving the IPC about a shared worker being terminated
and a new connection request received by the NetworkProcess. To address this,
I now added a CompletionHandler to the PostConnectEvent IPC from the
NetworkProcess to the SharedWorker process, indicating if we managed to connect
this SharedWorker object to the existing SharedWorker context. In case of
failure, we clean the state in the NetworkProcess and re-attempt the connection,
which will relaunch a SharedWorker.

* LayoutTests/http/wpt/shared-workers/relaunch-after-close-expected.txt: Added.
* LayoutTests/http/wpt/shared-workers/relaunch-after-close.html: Added.
* LayoutTests/http/wpt/shared-workers/resources/relaunch-after-close-page2.html: Added.
* LayoutTests/http/wpt/shared-workers/resources/relaunch-after-close-sharedworker.js: Added.
(shutdown):
(broadcast):
(sendMessageToPortRef):
* LayoutTests/http/wpt/shared-workers/resources/relaunch-after-close-window.js: Added.
(testComplete):
(async beginTestForPage1):
(async beginTestForPage2):
(async initSharedWorker):
(disonnectFromSharedWorker):
(async tellSharedWorkerToShutdown):
(log):
(deferredWithTimeout):
(async wait):
* Source/WebCore/workers/shared/context/SharedWorkerContextManager.cpp:
(WebCore::SharedWorkerContextManager::stopSharedWorker):
* Source/WebCore/workers/shared/context/SharedWorkerContextManager.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::sharedWorkerTerminated):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.cpp:
(WebKit::WebSharedWorkerServerToContextConnection::sharedWorkerTerminated):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::sharedWorkerTerminated):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.h:

Canonical link: https://commits.webkit.org/259228@main
